### PR TITLE
Chapter19

### PR DESCRIPTION
**Pull request by danieljharris**

### CODE REVIEW

File: Chapter19/19-1/db_tests.erl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+   ?_assertMatch({ok, 11},    read(10,    FilledTable) ),
+   ?_assertMatch({ok, 101},   read(100,   FilledTable) ),
+   ?_assertMatch({ok, 1001},  read(1000,  FilledTable) ),
+   ?_assertMatch({ok, 10001}, read(10000, FilledTable) ),
+   ?_assertMatch({error, key_not_found}, read(hello, FilledTable) )
+  ].
+
+
+match_write_test_() ->
+  {spawn,
```
Comment (velimir) [inline Chapter19/19-1/db_tests.erl]: Not sure what you need spawn here.
> {spawn, Tests}
>Runs the specified tests in a separate subprocess, while the current test process waits for it to finish. This is useful for tests that need a fresh, isolated process state. (Note that EUnit always starts at least one such a subprocess automatically; tests are never executed by the caller's own process.)

File: Chapter19/19-2/echoServer_tests.erl [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+     fun print_setup/0,     % setup function
+     fun print_cleanup/1,   % teardown function
+     fun print_test/1       % instantiator
+    }
+  }.
+
+print_setup() -> start().
+print_cleanup(_) -> stop().
+print_test(_) ->
+  [?_assertMatch({message, "Hello World"}, print("Hello World"))].
```
Comment (velimir) [inline Chapter19/19-2/echoServer_tests.erl]: it's better to use `assertEqual/2` if value on the left hand side is known. `assertEqual/2` prints out better report in that case if something has failed.

File: Chapter19/19-1/db_tests.erl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+-include_lib("eunit/include/eunit.hrl").
+
+new_test() ->
+  ?assertEqual( [] , new() ).
+
+destroy_test() ->
+  ?assertEqual( ok , destroy([{1,2},{3,4},{5,6}]) ).
+
+write_test_() ->
+  [?_assertMatch([{1,     element}], write(1,     element, []) ),
```
Comment (velimir) [inline Chapter19/19-1/db_tests.erl]: it's better to use assertEqual/2 if value on the left hand side is known. assertEqual/2 prints out better report in that case if something has failed.

File: Chapter19/19-3/db_server_tests.erl [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  start(),
+  write(1, 2),
+  write(3, 4),
+  code_upgrade().
+
+upgrade_cleanup(_) -> stop().
+
+%% Here you would check the db has been upgraded to the new format
+upgrade_test(NewDb) ->
+  [?_assertMatch([{3, 4}, {1, 2}], NewDb)].
```
Comment (velimir) [inline Chapter19/19-3/db_server_tests.erl]: How does it check code upgrade? As far as I can see format has to be changed from lists to dict (as it happened in chapter 8).

File: Chapter19/19-4/otp_testing_tests.erl [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                      add_usr/2]).
+
+% -define(NOTEST, true).
+-include_lib("eunit/include/eunit.hrl").
+
+path() -> "/Users/daniel.harris/Documents/GitErlang/learning-erlang/Chapter19/19-4/DB".
+
+
+%% Note -----------------------------------------------------------------
+% For some reason I can only test one thing at a time, otherwise they
```
Comment (velimir) [inline Chapter19/19-4/otp_testing_tests.erl]: make sure that processes that you spawned are stopped after each test. Or create a unit test, that includes all tests as a sequential list of commands, for instance: start_link, add_user, set_service, lookup_id, set_status, service_flag and then stop. All of those actions ideally should be tested.

File: Chapter19/19-5/my_db_gen_tests.erl [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+% -define(NOTEST, true).
+-include_lib("eunit/include/eunit.hrl").
+
+
+
+%% code:add_path("my_db-3.0/ebin/").
+
+
+
+start_link_test_() ->
```
Comment (velimir) [inline Chapter19/19-5/my_db_gen_tests.erl]: I suggest to no start and stop application on each test case, or if it's necessary use `foreach` fixture.

**Review by velimir — CHANGES_REQUESTED** [review]
can you please have a look at comments? Also you use spawn in all tests, which, really is not necessary, and, it seems to me, can be voided.

File: Chapter19/19-4/otp_testing_tests.erl (hunk 5 above)
Comment (danieljharris) [reply Chapter19/19-4/otp_testing_tests.erl]: Code now starts the server (as a test) runs other tests and stops the server (as a test) all in one go now, so nothing overlaps

File: Chapter19/19-5/my_db_gen_tests.erl (hunk 6 above)
Comment (danieljharris) [reply Chapter19/19-5/my_db_gen_tests.erl]: Combined all the tests in to one so the server only starts and stops once

**Review by velimir — APPROVED** [review]

**Pull request merged**